### PR TITLE
Add a license checker script

### LINKED ISSUE #6508: Ensure that all node modules used in production build are GPLv2+ compatible
With Guteneberg requiring almost 300 different node modules, we need to make sure anything that gets shipped to our end users is licensed in a GPL compatible way. 

Using [license-checker](https://github.com/davglass/license-checker) to look at all this:

```bash
$ license-checker --summary --production
├─ MIT: 221
├─ ISC: 35
├─ BSD-2-Clause: 6
├─ Apache-2.0: 5
├─ BSD-3-Clause: 5
├─ BSD: 3
├─ GPL-2.0+: 2
├─ Custom: https://s.w.org/style/images/codeispoetry.png: 2
├─ Apache License, Version 2.0: 2
├─ GPL-3.0: 1
├─ Custom: https://make.wordpress.org/design/handbook/design-guide/browser-support/: 1
├─ CC0-1.0: 1
├─ Custom: https://codex.wordpress.org/I18n: 1
├─ Custom: https://google.com: 1
├─ BSD*: 1
├─ GPL-2.0*: 1
├─ (WTFPL OR MIT): 1
├─ Custom: https://make.wordpress.org/accessibility/2015/04/15/let-wordpress-speak-new-in-wordpress-4-2/: 1
├─ (MIT AND CC-BY-3.0): 1
├─ Unlicense: 1
└─ (GPL-2.0 OR MIT): 1
```

The Apache-2.0 licensed isn't compatible with GPL v2, so we will need to find replacements for, convince them to update their license, or ensure they aren't included in the production build:

`doctrine@2.1.0`  Apache-2.0
`eslint-visitor-keys`  Apache-2.0
` mousetrap` Apache-2.0
`rx-lite`  Apache License, Version 2.0
`rx-lite-aggregates` Apache License, Version 2.0
`spdx-correct` Apache-2.0
`validate-npm-package-license`  Apache-2.0

Also, there is one GPL3 module 
` dom-react`  GPL-3.0

This isn't exhaustive and we likely need to go through each submodule by hand to ensure we don't inadvertently include something like code mirror's jshint file (which core did).

### PR DESCRIPTION
**Pull request by pento**
## Description

To ensure we're only adding GPL2 compatible packages, we need to check all the licenses.

This PR adds a script with defined safe licenses, and runs it in the `postinstall` script, to ensure devs are immediately aware if they've added something incompatible.

Due to the myriad of ways that package authors define their licenses, this script allows for license strings that aren't SPDX, and for when licenses aren't defined in the package's `package.json` but can be found in a `LICENSE` or `README` file.

Fixes #6508, #7822.

## Testing

`npm run check-licenses` will test that all packages defined in `dependencies` are GPL2 compatible, and that all `devDependencies` an OSS license.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code follows the accessibility standards.
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by gziolo** [comment]
It sort of works:

> [1]  ERROR  Module prompts has an incompatible license 'undefined'.
> [1]  ERROR  Module cyclist has an incompatible license 'undefined'.
> [1]  ERROR  Module cli-table has an incompatible license 'undefined'.

😅

**Comment by pento** [comment]
Good times. I sure do enjoy when things behave differently locally.

File: packages/scripts/scripts/check-licenses.js [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	const formattedAllowedLicense = allowedLicense.toLowerCase();
+	const formattedlicenseType = licenseType.toLowerCase();
+
+	if ( formattedAllowedLicense === formattedlicenseType ) {
+		return true;
+	}
+
+	if ( licenseType.indexOf( 'OR' ) < 0 ) {
```
Comment (gziolo) [inline packages/scripts/scripts/check-licenses.js]: Can we add an inline comment explaining what happens in here?

File: packages/scripts/scripts/check-licenses.js [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+		return false;
+	}
+
+	const subLicenseTypes = formattedlicenseType.replace( /^\(*/g, '' ).replace( /\)*$/, '' ).split( ' or ' ).map( ( e ) => e.trim() );
```
Comment (gziolo) [inline packages/scripts/scripts/check-licenses.js]: Same as above, I bet w are extracting sublicenses. However, it is not clear how it works.

**Review by gziolo — APPROVED** [review]
Love this PR ❤️  I'm happy that it integrates into `wp-scripts` so seamlessly and even takes advantage of `hasCliArg` util.

I think we could iterate on the usage of `reduce` functions which could stop processing when value is found like in the example I shared: https://github.com/WordPress/gutenberg/pull/8808#discussion_r209166052

As a general concept, it is good to go. I'm giving 👍 for the functionality.

**Comment by gziolo** [comment]
8bc21e9 - this is super helpful. Your future self will be very grateful for those comments :)

File: packages/scripts/scripts/check-licenses.js [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+ */
+const { hasCliArg } = require( '../utils' );
+
+const ERROR = chalk.reset.inverse.bold.red( ' ERROR ' );
+
+const prod = hasCliArg( '--prod' ) || hasCliArg( '--production' );
+const dev = hasCliArg( '--dev' ) || hasCliArg( '--development' );
+const gpl2 = hasCliArg( '--gpl2' );
+
+const gpl2Licenses = [
```
Comment (tofumatt) [inline packages/scripts/scripts/check-licenses.js]: Nitpick here, but it'd be nice to call this `gpl2CompatibleLicenses` or something like that.

File: packages/scripts/scripts/check-licenses.js [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+	'MPL-2.0',
+	'Public Domain',
+	'Unlicense',
+	'WTFPL',
+	'Zlib',
+	'(MIT AND BSD-3-Clause)',
+	'(MIT AND Zlib)',
+];
+
+const ossLicenses = [
```
Comment (tofumatt) [inline packages/scripts/scripts/check-licenses.js]: What does this variable mean? Aren't all of these licenses open source software licenses? 😄

File: packages/scripts/scripts/check-licenses.js [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+		'Permission is hereby granted, free of charge,',
+		'## License\n\nMIT',
+		'## License\n\n  MIT',
+	],
+};
+
+/**
+ * Check if a license string matches the given license.
+ *
+ * The license string can be a single license, or an SPDX-compatible OR license string.
```
Comment (tofumatt) [inline packages/scripts/scripts/check-licenses.js]: Grammar nitpick here but I think removing the comma (which isn't really needed for just a single "or" statement like this) would improve readability. I thought the next "OR" was another "or" in the sentence, but eventually realised "SPDX-compatible OR License String" was a thing. Maybe capitalising "License String" would make it clear that's a whole concept. I didn't know what "SPDX-compatible OR License String" was 😄

**Review by tofumatt — APPROVED** [review]
That's cool; I have a few little thoughts. Good idea 👍

File: packages/scripts/scripts/check-licenses.js (hunk 3 above)
Comment (pento) [reply packages/scripts/scripts/check-licenses.js]: 👍🏻

File: packages/scripts/scripts/check-licenses.js (hunk 5 above)
Comment (pento) [reply packages/scripts/scripts/check-licenses.js]: 👍🏻

File: packages/scripts/scripts/check-licenses.js (hunk 4 above)
Comment (pento) [reply packages/scripts/scripts/check-licenses.js]: THESE ARE NOT THE FREE-EST LICENSES MATT

RICHARD STALLMAN IS JUDGING US FOR EVEN MENTIONING THEM

😁

**Review by ntwb — APPROVED** [review]
:shipit:

File: packages/scripts/scripts/check-licenses.js (hunk 4 above)
Comment (tofumatt) [reply packages/scripts/scripts/check-licenses.js]: I CLEARLY HATE FREEDOM, MY MICROWAVE IS RUNNING APACHE 2 FIRMWARE.

**Pull request merged**